lld/ELF/MarkLive.cpp

```diff
@@ -201,45 +238,126 @@ void MarkLive<ELFT>::enqueue(InputSectionBase *sec, uint64_t offset) {
     return;
   sec->partition = sec->partition ? 1 : partition;
 
+  if (TrackWhyLive) {
+    if (sym) {
+      // If a specific symbol is referenced, that makes it alive. It may in turn
+      // make its section alive.
+      whyLive.try_emplace(sym, reason);
+      whyLive.try_emplace(sec, LiveReason{sym, "contained live symbol"});
+    } else {
+      // Otherwise, the reference generically makes the section live.
+      whyLive.try_emplace(sec, reason);
+    }
+  }
+
   // Add input section to the queue.
   if (InputSection *s = dyn_cast<InputSection>(sec))
     queue.push_back(s);
 }
 
-template <class ELFT> void MarkLive<ELFT>::markSymbol(Symbol *sym) {
+// Print the stack of reasons that the given symbol is live.
+template <class ELFT, bool TrackWhyLive>
+void MarkLive<ELFT, TrackWhyLive>::printWhyLive(Symbol *s) const {
+  // Skip dead symbols. A symbol is dead if it belongs to a dead section.
+  if (auto *d = dyn_cast<Defined>(s)) {
+    auto *sec = dyn_cast_or_null<InputSectionBase>(d->section);
+    if (sec && !sec->isLive())
+      return;
+  }
+
+  auto msg = Msg(ctx);
+
+  const auto printSymbol = [&](Symbol *s) {
+    if (s->isLocal())
```

Comment:
It may be worth printing the file for globals as well. It isn't needed to disambiguate the function, but it means I have to search in my source code for it if I want to find it.

Not a strong opinion at first as I expect we can introduce some form of user policy for how much information to print if it is requested by users.